### PR TITLE
Spring Cloud Zuul support

### LINKED ISSUE #929: Stormpath Zuul Spring Cloud Plugin
Allow the Stormpath Spring Boot Default Starter to be used as a plugin in a Spring Cloud Netflix Zuul deployment.  This allows a Zuul gateway to easily perform user management, auth, and security concerns via Stormpath so backend apps and microservices don't have to worry about that stuff as much or at all.

### PR DESCRIPTION
**Pull request by lhazlewood**
Fixes #929

Allows Stormpath to operate in a Zuul gateway so destination/origin servers do not need any Stormpath SDKs if desired.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -376,6 +399,37 @@
...
+                    </exclusion>
+                    <exclusion>
+                        <groupId>org.springframework.boot</groupId>
+                        <artifactId>spring-boot-starter-web</artifactId>
+                    </exclusion>
+                    <!--
+                    <exclusion>
+                        <groupId>org.springframework.boot</groupId>
+                        <artifactId>spring-boot-starter-actuator</artifactId>
+                    </exclusion> -->
```
Comment (mraible) [inline pom.xml]: I'd remove this commented-out dependency.

File: examples/zuul-spring-cloud-starter/pom.xml [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  ~ limitations under the License.
+  -->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>com.stormpath.sdk</groupId>
+        <artifactId>stormpath-sdk-examples</artifactId>
+        <version>1.0.4-SNAPSHOT</version>
```
Comment (jarias) [inline examples/zuul-spring-cloud-starter/pom.xml]: Version should be 1.1.0-SNAPSHOT

File: extensions/spring/cloud/pom.xml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  -->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>com.stormpath.spring</groupId>
+        <artifactId>stormpath-spring-parent</artifactId>
+        <version>1.0.4-SNAPSHOT</version>
```
Comment (jarias) [inline extensions/spring/cloud/pom.xml]: Update version

File: extensions/spring/cloud/stormpath-zuul-spring-cloud-starter/pom.xml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>com.stormpath.spring</groupId>
+        <artifactId>stormpath-spring-parent</artifactId>
+        <version>1.0.4-SNAPSHOT</version>
```
Comment (jarias) [inline extensions/spring/cloud/stormpath-zuul-spring-cloud-starter/pom.xml]: Bump version

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ResourceToMapConverter.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        }
+
+        if (getExcludedFields().contains(propName)) {
+            return false;
+        }
+
+        return !(propValue instanceof Resource) || propValue instanceof CustomData;
+    }
+
+    public Map<String, Object> toMap(Object r) {
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ResourceToMapConverter.java]: Argument `Object r` should be `Object resource`. The assertion explains "resource cannot be null".

Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ResourceToMapConverter.java]: This generic function seems to be quite useful. It could be moved to a `ResourceUtils` class in the core SDK so it can be re-used by other modules in the future.

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ResourceToMapConverter.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+            }
+
+            props.put(propName, propValue);
+        }
+
+        return props;
+    }
+
+    @SuppressWarnings("unchecked")
+    private Map<String, Object> deepCopy(Map<String, Object> src) {
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ResourceToMapConverter.java]: This `deepCopy` function could be made public in the proposed `ResourceUtils` class as it also has some potential to be reusable in the future.

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ResourceToMapConverter.java [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+                value = deepCopy((Collection) value);
+            }
+            dst.put(name, value);
+        }
+
+        return dst;
+    }
+
+    @SuppressWarnings("unchecked")
+    private Collection deepCopy(Collection src) {
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ResourceToMapConverter.java]: Idem

File: extensions/spring/cloud/pom.xml [hunk 8]
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Copyright 2015 Stormpath, Inc.
```
Comment (mrioan) [inline extensions/spring/cloud/pom.xml]: 2016

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ResourceToMapConverter.java [hunk 9]
```diff
@@ -0,0 +1,189 @@
+package com.stormpath.sdk.servlet.mvc;
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ResourceToMapConverter.java]: Missing header

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/json/ResourceJsonFunction.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
+package com.stormpath.sdk.servlet.json;
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/json/ResourceJsonFunction.java]: Missing header

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/json/JsonFunction.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
+package com.stormpath.sdk.servlet.json;
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/json/JsonFunction.java]: Missing header

File: examples/zuul-spring-cloud-starter/src/main/java/com/stormpath/spring/cloud/examples/SecurityConfig.java [hunk 12]
```diff
@@ -0,0 +1,17 @@
+package com.stormpath.spring.cloud.examples;
```
Comment (mrioan) [inline examples/zuul-spring-cloud-starter/src/main/java/com/stormpath/spring/cloud/examples/SecurityConfig.java]: Missing header

File: examples/zuul-spring-cloud-starter/src/main/java/com/stormpath/spring/cloud/examples/Application.java [hunk 13]
```diff
@@ -0,0 +1,11 @@
+package com.stormpath.spring.cloud.examples;
```
Comment (mrioan) [inline examples/zuul-spring-cloud-starter/src/main/java/com/stormpath/spring/cloud/examples/Application.java]: Missing header

File: api/src/main/java/com/stormpath/sdk/lang/Function.java [hunk 14]
```diff
@@ -0,0 +1,25 @@
+package com.stormpath.sdk.lang;
```
Comment (mrioan) [inline api/src/main/java/com/stormpath/sdk/lang/Function.java]: Missing header

File: extensions/spring/cloud/stormpath-zuul-spring-cloud-starter/pom.xml [hunk 15]
```diff
@@ -0,0 +1,45 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Copyright 2015 Stormpath, Inc.
```
Comment (mrioan) [inline extensions/spring/cloud/stormpath-zuul-spring-cloud-starter/pom.xml]: 2016

File: extensions/spring/cloud/stormpath-zuul-spring-cloud-starter/src/main/java/com/stormpath/spring/cloud/zuul/autoconfigure/StormpathZuulAutoConfiguration.java [hunk 16]
```diff
@@ -0,0 +1,103 @@
+package com.stormpath.spring.cloud.zuul.autoconfigure;
```
Comment (mrioan) [inline extensions/spring/cloud/stormpath-zuul-spring-cloud-starter/src/main/java/com/stormpath/spring/cloud/zuul/autoconfigure/StormpathZuulAutoConfiguration.java]: Header

File: extensions/zuul/pom.xml [hunk 17]
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Copyright 2014 Stormpath, Inc.
```
Comment (mrioan) [inline extensions/zuul/pom.xml]: 2016

File: extensions/zuul/src/main/java/com/stormpath/zuul/account/DefaultAccountHeaderValueResolver.java [hunk 18]
```diff
@@ -0,0 +1,101 @@
+package com.stormpath.zuul.account;
```
Comment (mrioan) [inline extensions/zuul/src/main/java/com/stormpath/zuul/account/DefaultAccountHeaderValueResolver.java]: Header

File: extensions/zuul/src/main/java/com/stormpath/zuul/account/DefaultAccountHeaderValueResolver.java [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+ * If an account is located, the default configuration returns a JSON representation of the account.  A
+ * different account String representation can be returned by configuring a different
+ * {@link #setAccountStringFactory(Function) accountStringFactory}.
+ * </p>
+ *
+ * @see #setAccountResolver(AccountResolver)
+ * @see #setAccountStringFactory(Function)
+ * @since 1.1.0
+ */
+public class DefaultAccountHeaderValueResolver implements Resolver<String> {
```
Comment (mrioan) [inline extensions/zuul/src/main/java/com/stormpath/zuul/account/DefaultAccountHeaderValueResolver.java]: Why the name contains the word `Header`? Shouldn't it be something like `DefaultAccountValueResolver`?

File: extensions/zuul/src/main/java/com/stormpath/zuul/account/ForwardedAccountHeaderFilter.java [hunk 20]
```diff
@@ -0,0 +1,89 @@
+package com.stormpath.zuul.account;
```
Comment (mrioan) [inline extensions/zuul/src/main/java/com/stormpath/zuul/account/ForwardedAccountHeaderFilter.java]: Header

File: extensions/zuul/src/main/java/com/stormpath/zuul/account/ForwardedAccountHeaderFilter.java [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+/**
+ * An Zuul {@code pre} filter that intercepts requests and sets an outbound request header with a string representation
+ * of the account currently associated with the request.  This allows destination/origin http servers to read the
+ * request account's information by simply looking at a request header, without needing any Stormpath SDK if desired.
+ * <p>
+ * This filter will only execute if an account is available as determined by invoking its
+ * {@link #setAccountResolver(AccountResolver) accountResolver}'s
+ * {@link AccountResolver#hasAccount(ServletRequest) hasAccount} method. If there is an account, the account's string
+ * representation will be obtained by calling the {@link #valueResolver}, a
+ * {@link DefaultAccountHeaderValueResolver} instance by default.  If there is not an account associated with the
```
Comment (mrioan) [inline extensions/zuul/src/main/java/com/stormpath/zuul/account/ForwardedAccountHeaderFilter.java]: Update here if `DefaultAccountHeaderValueResolver` is actually renamed as suggested above

File: extensions/zuul/src/main/java/com/stormpath/zuul/account/ForwardedAccountHeaderFilter.java [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    /**
+     * Sets the {@code AccountResolver} used to determine if this filter should execute during a request or not.  The
+     * default instance is a {@link DefaultAccountResolver}.
+     *
+     * @param accountResolver the {@code AccountResolver} used to determine if this filter should execute during a request or not.
+     */
+    public void setAccountResolver(AccountResolver accountResolver) {
+        this.accountResolver = accountResolver;
```
Comment (mrioan) [inline extensions/zuul/src/main/java/com/stormpath/zuul/account/ForwardedAccountHeaderFilter.java]: Assert.notNull ?

**Comment by mrioan** [comment]
@lhazlewood I provided a few comments.
Additionally, I do not see any tests. You could offload yourself from writing them since anyone in the team could do that. However it would still be nice if you could suggest some important things that should be tested for this gateway. Thanks!

**Comment by mrioan** [comment]
@lhazlewood I provided a few comments.
Additionally, I do not see any tests. You could offload yourself from writing them since anyone in the team could do that. However it would still be nice if you could suggest some important things that should be tested for this gateway. Thanks!

File: examples/zuul-spring-cloud-starter/pom.xml (hunk 2 above)
Comment (lhazlewood) [inline examples/zuul-spring-cloud-starter/pom.xml]: I can't until the 1.1.x branch does the same - merge conflicts.

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ResourceToMapConverter.java (hunk 5 above)
Comment (lhazlewood) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ResourceToMapConverter.java]: I chose to use the variable named `resource` once the object was known to be an `AbstractResource`.  To that end, I renamed this to `object` so that it is more clear.  I'll update the PR shortly.  Thanks!

Comment (lhazlewood) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ResourceToMapConverter.java]: I'm glad to see you saw the similarities to `DefaultAccountModelFactory` @mrioan ;)  It was indeed my intention to use the new (generic) `ResourceToMapConverter` as a full replacement of the `DefaultAccountModelFactory` implementation - the former is more configurable and is 100% generic with no ties to MVC concerns.  I didn't want to do that as part of this pull request however as I see that refactoring as a different scope of work.  I just opened #961 to track this work. Thanks!

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ResourceToMapConverter.java (hunk 6 above)
Comment (lhazlewood) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/ResourceToMapConverter.java]: I'd probably do that later if/when we refactor ResourceUtils.  No need to expose the visibility for this particular pull request.  If you'd like to see that refactoring, can you please open a new issue?

File: extensions/spring/cloud/pom.xml (hunk 3 above)
Comment (lhazlewood) [inline extensions/spring/cloud/pom.xml]: can't until 1.1.x source branch does.

File: extensions/spring/cloud/stormpath-zuul-spring-cloud-starter/pom.xml (hunk 4 above)
Comment (lhazlewood) [inline extensions/spring/cloud/stormpath-zuul-spring-cloud-starter/pom.xml]: same

File: extensions/zuul/src/main/java/com/stormpath/zuul/account/DefaultAccountHeaderValueResolver.java (hunk 19 above)
Comment (lhazlewood) [inline extensions/zuul/src/main/java/com/stormpath/zuul/account/DefaultAccountHeaderValueResolver.java]: It was intended to reflect that the value returned would be used as an HTTP header value.

That said, after a few refactorings, this particular implementation doesn't need to 'know' about headers at all - it is more of a delegate resolver that converts to a string, and headers aren't a concern at all.  I'll change the name to something that better reflects the minimal scope of the implementation.  Good catch!

File: extensions/zuul/src/main/java/com/stormpath/zuul/account/ForwardedAccountHeaderFilter.java (hunk 21 above)
Comment (lhazlewood) [inline extensions/zuul/src/main/java/com/stormpath/zuul/account/ForwardedAccountHeaderFilter.java]: Ack.  Thanks!

File: extensions/zuul/src/main/java/com/stormpath/zuul/account/ForwardedAccountHeaderFilter.java (hunk 22 above)
Comment (lhazlewood) [inline extensions/zuul/src/main/java/com/stormpath/zuul/account/ForwardedAccountHeaderFilter.java]: 👍

File: impl/src/main/java/com/stormpath/sdk/impl/http/support/UserAgent.java [hunk 23]
```diff
@@ -63,6 +63,10 @@
     //SDK
     private static final String STORMPATH_SDK_STRING = "stormpath-sdk-java";
 
+    //Stormpath Zuul support
+    private static final String STORMPATH_ZUUL_ID = "stormpath-zuul";
+    private static final String STORMPATH_ZUUL_CLASS = "com.stormpath.zuul.account.AccountFilter";
```
Comment (lhazlewood) [inline impl/src/main/java/com/stormpath/sdk/impl/http/support/UserAgent.java]: Note to self - this class does not exist.  Need to update it to reflect the correct class name.

**Comment by lhazlewood** [comment]
@mrioan PR updated with tests!  Can you please re-review?

File: impl/src/main/java/com/stormpath/sdk/impl/authc/BasicApiAuthenticator.java [hunk 24]
```diff
@@ -56,24 +52,7 @@ public ApiAuthenticationResult authenticate(Application application, DefaultBasi
...
-                    throw new RuntimeException(e);
-                }
-                try {
-                    Thread.sleep(500);
-                } catch (InterruptedException e1) {
-                    throw new RuntimeException(e1);
-                }
-            }
-        }
+        ApiKey apiKey = application.getApiKey(id, new DefaultApiKeyOptions().withAccount());
```
Comment (mrioan) [inline impl/src/main/java/com/stormpath/sdk/impl/authc/BasicApiAuthenticator.java]: @lhazlewood Why was this removed? The comment says that this loop is a workaround for #440 . Is that issue known to have disappeared?

File: impl/src/test/groovy/com/stormpath/sdk/impl/authc/BasicApiAuthenticatorTest.groovy [hunk 25]
```diff
@@ -1,101 +0,0 @@
-package com.stormpath.sdk.impl.authc
```
Comment (mrioan) [inline impl/src/test/groovy/com/stormpath/sdk/impl/authc/BasicApiAuthenticatorTest.groovy]: @lhazlewood why was this removed?

**Comment by mrioan** [comment]
@lhazlewood I provided 2 comments. Additionally, what is your idea about documentation? The gateway will not be simple to use unless we have some documentation. Maybe you can add a short quickstart right in the example (in the examples) folder so early adopters can start to play with it before we can create the complete documentation. Back to you. Thanks!

File: impl/src/main/java/com/stormpath/sdk/impl/authc/BasicApiAuthenticator.java (hunk 24 above)
Comment (lhazlewood) [inline impl/src/main/java/com/stormpath/sdk/impl/authc/BasicApiAuthenticator.java]: I didn't touch this - I think it was pulled in from a merge to 1.1.x or master?

File: impl/src/test/groovy/com/stormpath/sdk/impl/authc/BasicApiAuthenticatorTest.groovy (hunk 25 above)
Comment (lhazlewood) [inline impl/src/test/groovy/com/stormpath/sdk/impl/authc/BasicApiAuthenticatorTest.groovy]: Didn't touch this either - came in from a merge to master or 1.1.x?

File: impl/src/main/java/com/stormpath/sdk/impl/authc/BasicApiAuthenticator.java (hunk 24 above)
Comment (lhazlewood) [inline impl/src/main/java/com/stormpath/sdk/impl/authc/BasicApiAuthenticator.java]: I'm confused as to why this modification is in this PR - it must have come from master?

File: impl/src/test/groovy/com/stormpath/sdk/impl/authc/BasicApiAuthenticatorTest.groovy (hunk 25 above)
Comment (lhazlewood) [inline impl/src/test/groovy/com/stormpath/sdk/impl/authc/BasicApiAuthenticatorTest.groovy]: Fixing and restoring...

File: impl/src/main/java/com/stormpath/sdk/impl/authc/BasicApiAuthenticator.java (hunk 24 above)
Comment (lhazlewood) [inline impl/src/main/java/com/stormpath/sdk/impl/authc/BasicApiAuthenticator.java]: Fixing and restoring...

**Comment by lhazlewood** [comment]
Fixed those two issues @mrioan - I was expecting to do docs as part of the doc merge effort - there's a lot to cover, so a simple quickstart might not be enough.  I'll see what I can do today...

**Comment by mrioan** [comment]
@lhazlewood It all looks good to me now. Feel free to merge when Travis passes. woo–hoo! Gateway! Awesome! Great work @lhazlewood
